### PR TITLE
Robotarm

### PR DESCRIPTION
**Pull request by finleylau**

### CODE REVIEW

File: client/src/components/ActiveStimulus/ContinuousWorld/index.js [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+export default class ContinuousWorld extends React.Component {
+    constructor(props) {
+        super(props)
+
+        var data = this.props.data
+
+        this.robotSpecs = {
+            width: data.width,
+            height: data.height,
+            velocity: data.velocity, // in pixels/second
```
Comment (deepakgopinath) [inline client/src/components/ActiveStimulus/ContinuousWorld/index.js:15]: is this the width and height of the window?

File: client/src/components/ActiveStimulus/ContinuousWorld/index.js [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+        }
+
+        this.robotHalfDiagonal = 0.5 * this.robotSpecs.width * Math.sqrt(2)
+
+        this.goalSpecs = {
+            width: data.goalWidth,
+            height: data.goalHeight,
+            x: data.goalLocationX,
+            y: data.goalLocationY
```
Comment (deepakgopinath) [inline client/src/components/ActiveStimulus/ContinuousWorld/index.js:26]: is there a reason why there is not orientation field for the goal?

File: client/src/components/ActiveStimulus/components/RobotArm/RobotArm2.js [hunk 3]
```diff
@@ -0,0 +1,654 @@
...
+import Matter, { Events, Composite, World, Body } from "matter-js";
+import { Button } from 'react-bootstrap'
+
+class Scene extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            currentNode: 0,
+            mode: 0 // 0 controls nodes, 1 controls hand angle, 2 controls hand pos
+        };
```
Comment (deepakgopinath) [inline client/src/components/ActiveStimulus/components/RobotArm/RobotArm2.js:12]: We probably need to have 2 modes: 1) Joint control and 2) Cartesian control. Where joint control is essentially (0 and 1) and cartesian is (1 and 2)

File: client/src/components/ActiveStimulus/components/RobotArm/RobotArm2.js [hunk 4]
```diff
@@ -0,0 +1,654 @@
...
+            Body.setStatic(i, true)
+        }
+
+        World.add(this.world, this.walls)
+
+        this.linkHeight = 100
+        var linkWidth = 25
+        var jointOffset = 5
+        var linkRadius = 10
+        var constraintsVisible = true
```
Comment (deepakgopinath) [inline client/src/components/ActiveStimulus/components/RobotArm/RobotArm2.js:90]: Could we possibly not have these dimensions hard coded?

File: client/src/components/ActiveStimulus/components/RobotArm/RobotArm2.js [hunk 5]
```diff
@@ -0,0 +1,654 @@
...
+            {
+                color: "blue"
+            },
+            {
+                color: "orange"
+            },
+            {
+                color: "purple"
+            }
+        ]
```
Comment (deepakgopinath) [inline client/src/components/ActiveStimulus/components/RobotArm/RobotArm2.js:105]: Maybe have a default color if someone wants to do more than 4 links!! Very unlikely we need a 5 degree of freedom robotic arm.

File: client/src/components/ActiveStimulus/ContinuousWorld/index.js [hunk 6]
```diff
@@ -0,0 +1,317 @@
+import React from 'react'
+import TwoDWorld from "../components/TwoDWorld"
+import { Row, Col } from 'react-bootstrap'
+var SAT = require('sat');
+
+export default class ContinuousWorld extends React.Component {
+    constructor(props) {
+        super(props)
```
Comment (deepakgopinath) [inline client/src/components/ActiveStimulus/ContinuousWorld/index.js:8]: could we add documentation regarding what this world is?

File: client/src/components/ActiveStimulus/ContinuousWorld/index.js [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+
+        var data = this.props.data
+
+        this.robotSpecs = {
+            width: data.width,
+            height: data.height,
+            velocity: data.velocity, // in pixels/second
+            angularVelocity: data.angularVelocity
+
```
Comment (deepakgopinath) [inline client/src/components/ActiveStimulus/ContinuousWorld/index.js:17]: this, I believe is the wheelchair? In general, more comments are always welcome/needed.

File: client/src/components/ActiveStimulus/ContinuousWorld/index.js [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+        this.lastRender = 0
+        this.didWin = false
+        this.then = Date.now()
+        this.startTime = this.then
+        this.start = Date.now()
+
+        this.eventLog = []
+
+        this.fpsInterval = 1000 / data.fps
+
```
Comment (deepakgopinath) [inline client/src/components/ActiveStimulus/ContinuousWorld/index.js:59]: More comments....

File: client/src/components/ActiveStimulus/ContinuousWorld/index.js [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+        } else {
+            this.eventLog.push({
+                keys: keysPressed,
+                time: time,
+                lastUpdated: time
+            })
+        }
+
+        // progress is the amount of time that has passed in ms
+        var robotVelocity = this.robotSpecs.velocity / 60 // pixels/frame
```
Comment (deepakgopinath) [inline client/src/components/ActiveStimulus/ContinuousWorld/index.js:181]: what is 60 in these two lines? Should that be hardcoded?

File: client/src/components/ActiveStimulus/ContinuousWorld/index.js [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+    componentWillUnmount() {
+        document.removeEventListener("keydown", this.handleKeyPress, false);
+        document.removeEventListener("keyup", this.handleKeyPress, false);
+    }
+
+
+
+    handleKeyPress(event) {
+        var validKeys = new Set(["w", "a", "s", "d", "ArrowUp", "ArrowLeft", "ArrowRight", "ArrowDown"])
+        if (validKeys.has(event.key)) this.keys[event.key] = event.type === "keydown"
```
Comment (deepakgopinath) [inline client/src/components/ActiveStimulus/ContinuousWorld/index.js:263]: Can this handle uppercase W,A,S,D as well?

File: client/src/components/ActiveStimulus/components/RobotArm/RobotArm2.js [hunk 11]
```diff
@@ -0,0 +1,654 @@
...
+            Runner = Matter.Runner,
+            Body = Matter.Body,
+            Composite = Matter.Composite,
+            Composites = Matter.Composites,
+            Constraint = Matter.Constraint,
+            MouseConstraint = Matter.MouseConstraint,
+            Mouse = Matter.Mouse,
+            World = Matter.World,
+            Bodies = Matter.Bodies;
+
```
Comment (deepakgopinath) [inline client/src/components/ActiveStimulus/components/RobotArm/RobotArm2.js:49]: Physics Engine stuff? If so, more comments.

File: client/src/components/ActiveStimulus/components/RobotArm/RobotArm2.js [hunk 12]
```diff
@@ -0,0 +1,654 @@
...
+                bodyB: this.links[i],
+                length: 0,
+                pointA: { x: 0, y: prevOffset },
+                pointB: { x: 0, y: this.linkHeight / 2 + jointOffset },
+                stiffness: 1
+            }))
+        }
+
+        this.centerSupports = []
+        this.btSupports = []
```
Comment (deepakgopinath) [inline client/src/components/ActiveStimulus/components/RobotArm/RobotArm2.js:141]: Comments regarding what is happening in (140-220)...

File: client/src/components/ActiveStimulus/RobotArmWorld/index.js [hunk 13]
```diff
@@ -85,6 +83,8 @@ class Scene extends React.Component {
             this.goalY = json.goal.y
         }
 
+        this.jointOffset = 3
```
Comment (deepakgopinath) [inline client/src/components/ActiveStimulus/RobotArmWorld/index.js:86]: could you maybe leave a comment regarding what this variable is about?

File: client/src/components/ActiveStimulus/ContinuousWorld/index.js [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+            xd = robotVelocity * Math.cos(tf)
+            yd = robotVelocity * Math.sin(tf)
+        } else if (this.keys["s"] || this.keys["ArrowDown"]) {
+            xd = robotVelocity * -1 * Math.cos(tf)
+            yd = robotVelocity * -1 * Math.sin(tf)
+        }
+
+        var xf = xi + xd
+        var yf = yi + yd
+
```
Comment (deepakgopinath) [inline client/src/components/ActiveStimulus/ContinuousWorld/index.js:211]: could you please leave a comment saying this is where the kinematic update happens. Later on, potentially there are situations where we want to incorporate modifications to this kinematic update (by adding a noise term or a drift term, simulating different friction on the surfaces etc).

File: client/src/components/ActiveStimulus/ContinuousWorld/index.js [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+        //this.robot.rotate(td)
+        this.robot.setAngle(tf)
+        this.robot.pos.x = xf
+        this.robot.pos.y = yf
+
+
+        // assume we don't collide to start
+        var collisionObstacle = false
+        for (var o in this.obstacles) {
```
Comment (deepakgopinath) [inline client/src/components/ActiveStimulus/ContinuousWorld/index.js:221]: Is there an assert that can be incorporated to ensure that the starting position is not in collision?

File: client/src/Content.js [hunk 16]
```diff
@@ -197,6 +200,14 @@ export default class Content extends React.Component {
                         return (
                             <VideoStimulus key={this.state.trialData.trialIndex} data={this.state.trialData} submit={this.getNextTrial} />
                         )
+                    } else if (this.state.trialData.stimulusType === "continuous-world") {
+                        return (
+                            <ContinuousWorld key={this.state.trialData.trialIndex} data={this.state.trialData} submit={this.getNextTrial} />
```
Comment (deepakgopinath) [inline client/src/Content.js:205]: How hard would it be to change ContinuousWorld to something like WheelchairNav or something? Coz, the continuous world doesn't fully capture the fact that it is mobile navigation we are dealing with. You could have a robotic arm in a continuous world as well.

File: client/src/components/ActiveStimulus/ContinuousWorld/index.js [hunk 17]
```diff
@@ -0,0 +1,317 @@
...
+            width: data.width,
+            height: data.height,
+            velocity: data.velocity, // in pixels/second
+            angularVelocity: data.angularVelocity
+
+        }
+
+        this.robotHalfDiagonal = 0.5 * this.robotSpecs.width * Math.sqrt(2)
+
+        this.goalSpecs = {
```
Comment (deepakgopinath) [inline client/src/components/ActiveStimulus/ContinuousWorld/index.js:22]: Does this easily extend to multiple goal locations? I could see a scenario where we could have a study to test biases in how humans pick a goal from multiple options

File: client/src/components/ActiveStimulus/ContinuousWorld/index.js (hunk 10 above)
Comment (finleylau) [reply client/src/components/ActiveStimulus/ContinuousWorld/index.js:263]: uppercase W,A,S,D would not be able to be input

**Pull request closed without merging**